extensions/charts/vsphereoutoftree.go

```diff
@@ -0,0 +1,178 @@
+package charts
+
+import (
+	"github.com/rancher/shepherd/extensions/projects"
+	"github.com/rancher/shepherd/extensions/rke1/nodetemplates"
+	"github.com/rancher/shepherd/pkg/api/steve/catalog/types"
+
+	"github.com/rancher/shepherd/clients/rancher"
+	"github.com/rancher/shepherd/clients/rancher/catalog"
+	r1vsphere "github.com/rancher/shepherd/extensions/rke1/nodetemplates/vsphere"
+)
+
+const (
+	systemProject       = "System"
+	vsphereCPIchartName = "rancher-vsphere-cpi"
+	vsphereCSIchartName = "rancher-vsphere-csi"
+
+	vcenter      = "vCenter"
+	storageclass = "storageClass"
+
+	datacenters  = "datacenters"
+	host         = "host"
+	password     = "password"
+	username     = "username"
+	port         = "port"
+	clusterid    = "clusterId"
+	datastoreurl = "datastoreURL"
+)
+
+// InstallVsphereOutOfTreeCharts installs the CPI and CSI chart for aws cloud provider in a given cluster.
+func InstallVsphereOutOfTreeCharts(client *rancher.Client, vsphereTemplate *nodetemplates.NodeTemplate, repoName, clusterID string) error {
+
+	serverSetting, err := client.Management.Setting.ByID(serverURLSettingID)
+	if err != nil {
+		return err
+	}
+
+	registrySetting, err := client.Management.Setting.ByID(defaultRegistrySettingID)
+	if err != nil {
+		return err
+	}
+
+	project, err := projects.GetProjectByName(client, clusterID, systemProject)
+	if err != nil {
+		return err
+	}
+
+	catalogClient, err := client.GetClusterCatalogClient(clusterID)
+	if err != nil {
+		return err
+	}
+
+	latestCPIVersion, err := catalogClient.GetLatestChartVersion(vsphereCPIchartName, catalog.RancherChartRepo)
+	if err != nil {
+		return err
+	}
+
+	installCPIOptions := &InstallOptions{
+		ClusterID: clusterID,
+		Version:   latestCPIVersion,
+		ProjectID: project.ID,
+	}
+
+	chartInstallActionPayload := &payloadOpts{
+		InstallOptions:  *installCPIOptions,
+		Name:            vsphereCPIchartName,
+		Namespace:       kubeSystemNamespace,
+		Host:            serverSetting.Value,
+		DefaultRegistry: registrySetting.Value,
+	}
+
+	chartInstallAction, err := vsphereCPIChartInstallAction(catalogClient,
+		chartInstallActionPayload, vsphereTemplate, installCPIOptions, repoName, kubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+
+	err = catalogClient.InstallChart(chartInstallAction, repoName)
+	if err != nil {
+		return err
+	}
```

Comment:
This one doesn't need any cleanup registration correct? 

Cause basically if these charts are uninstalled the cluster won't be functional and there's no case where it's out-of-tree and without these charts right?

Reply:
correct